### PR TITLE
Add LR as tensor tests

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123753
* #123752
* #123751
* __->__ #123750

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by janeyx99 — COMMENTED** [review]
This change contains code that would immortalize the need for `_stack_if_compiling` in noncapturable code which breaks the invariant we had agreed upon before. The goal I had envisioned was to remove these from that codepath, but this PR moves in the opposite direction of engraining it.

**Comment by mlazos** [comment]
> This change contains code that would immortalize the need for `_stack_if_compiling` in noncapturable code which breaks the invariant we had agreed upon before. The goal I had envisioned was to remove these from that codepath, but this PR moves in the opposite direction of engraining it.

So this mostly exposed the bug that NAdam doesn't have this. If we are able to only take the capturable path (pending the other discussion) we can remove `_stack_if_compiling` from all of the optims.

**Comment by mlazos** [comment]
cc @janeyx99 this is ready again for review

File: torch/optim/sgd.py [hunk 1]
```diff
@@ -427,7 +427,12 @@ def _multi_tensor_sgd(
                 device_grads = bufs
 
         if not device_has_sparse_grad:
-            torch._foreach_add_(device_params, device_grads, alpha=-lr)
+            # handle internal item() call if lr is a tensor
+            if isinstance(lr, torch.Tensor) and torch._utils.is_compiling():
```
Comment (janeyx99) [inline torch/optim/sgd.py:431]: Everything else looks good, but I don’t like adding more is_compiling()s into the code.

What is the plan for compile to support scalar tensors like this? Are there alternatives where is_compiling wouldn’t be necessary?

Comment (mlazos) [reply torch/optim/sgd.py:431]: Dynamo handles scalar tensors already, the issue is that passing a scalar as an alpha or value kwarg in some torch functions results in an item() call internally. Symints are the long term plan there, but they aren't ready. Unfortunately I don't think there is an alternative here.

**Review by janeyx99 — APPROVED** [review]
meh, i like tests tho

**Pull request closed without merging**